File: typescript/src/keycheck.ts
/**
 * Ed25519 public-key validation for the verifier's trust boundary.
 *
 * Byte-for-byte mirror of the Python algovoi_rfc9421_verifier.keycheck module.
 * Rejects, fail-closed and BEFORE signature verification:
 *   1. non-canonical encodings (y >= p) and points not on the curve;
 *   2. small-order points (order dividing the cofactor 8, incl. the identity).
 *
 * @noble/ed25519 under its default ZIP215 rules accepts small-order public keys;
 * for an audited offline verifier that treats the resolved key as the trust
 * anchor we reject them. Small-order-ness is derived mathematically
 * ([8]P == identity) using RFC 8032 Appendix A arithmetic — no hard-coded
 * blocklist.
 */

const P = 2n ** 255n - 19n;

function mod(a: bigint): bigint {
  const r = a % P;
  return r >= 0n ? r : r + P;
}

function modPow(base: bigint, exp: bigint, m: bigint): bigint {
  let result = 1n;
  let b = base % m;
  if (b < 0n) b += m;
  let e = exp;
  while (e > 0n) {
    if (e & 1n) result = (result * b) % m;
    b = (b * b) % m;
    e >>= 1n;
  }
  return result;
}

function modInv(a: bigint): bigint {
  return modPow(mod(a), P - 2n, P);
}

const D = mod(-121665n * modInv(121666n));
const SQRT_M1 = modPow(2n, (P - 1n) / 4n, P);

// Extended homogeneous coordinates (X, Y, Z, T).
type Point = [bigint, bigint, bigint, bigint];

export class WeakKeyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WeakKeyError";
  }
}

function recoverX(y: bigint, sign: number): bigint | null {
  if (y >= P) return null; // non-canonical y encoding
  const x2 = mod((y * y - 1n) * modInv(mod(D * y * y + 1n)));
  if (x2 === 0n) return sign ? null : 0n;
  let x = modPow(x2, (P + 3n) / 8n, P);
  if (mod(x * x - x2) !== 0n) x = mod(x * SQRT_M1);
  if (mod(x * x - x2) !== 0n) return null; // not a square -> off curve
  if (Number(x & 1n) !== sign) x = P - x;
  return x;
}

function decode(pk: Uint8Array): Point | null {
  if (pk.length !== 32) return null;
  let y = 0n;
  for (let i = 31; i >= 0; i--) y = (y << 8n) | BigInt(pk[i]); // little-endian
  const sign = Number((y >> 255n) & 1n);
  y &= (1n << 255n) - 1n;
  const x = recoverX(y, sign);
  if (x === null) return null;
  return [x, y, 1n, mod(x * y)];
}

function add(p: Point, q: Point): Point {
  const A = mod((p[1] - p[0]) * (q[1] - q[0]));
  const B = mod((p[1] + p[0]) * (q[1] + q[0]));
  const C = mod(2n * p[3] * q[3] * D);
  const Dd = mod(2n * p[2] * q[2]);
  const E = B - A;
  const F = Dd - C;
  const G = Dd + C;
  const H = B + A;
  return [mod(E * F), mod(G * H), mod(F * G), mod(E * H)];
}

function mul8(p: Point): Point {
  const p2 = add(p, p);
  const p4 = add(p2, p2);
  return add(p4, p4);
}

function isIdentity(p: Point): boolean {
  return mod(p[0]) === 0n && mod(p[1] - p[2]) === 0n;
}

/** True iff pk decodes to a point whose order divides 8 (incl. identity). */
export function isSmallOrder(pk: Uint8Array): boolean {
  const p = decode(pk);
  if (p === null) return false;
  return isIdentity(mul8(p));
}

/** Throw WeakKeyError if pk is non-canonical, off-curve, or small-order. */
export function checkEd25519PublicKey(pk: Uint8Array): void {
  const p = decode(pk);
  if (p === null) {
    throw new WeakKeyError("public key is not a canonical on-curve Ed25519 point");
  }
  if (isIdentity(mul8(p))) {
    throw new WeakKeyError("public key is a small-order Ed25519 point");
  }
}

File: python/algovoi_rfc9421_verifier/keycheck.py
"""Ed25519 public-key validation for the verifier's trust boundary.

Rejects, fail-closed and *before* signature verification:

  1. non-canonical encodings  (y-coordinate >= p, i.e. a duplicate encoding of a
     valid point -- a malleability class), and points not on the curve;
  2. small-order points        (order dividing the cofactor 8, incl. the identity),
     which permit signature-malleability / cross-key verification classes.

Why this is needed: PyNaCl / libsodium's basic `crypto_sign_verify` and
@noble/ed25519 under its default ZIP215 rules both ACCEPT small-order (and some
non-canonical) public keys. For an audited offline verifier that treats the
resolved key as the trust anchor, accepting such a key is a defect. This gate is
self-contained (RFC 8032 Appendix A arithmetic, extended homogeneous coords) and
derives small-order-ness mathematically ([8]P == identity) rather than trusting a
hard-coded blocklist.
"""
from __future__ import annotations

_p = 2**255 - 19
_d = (-121665 * pow(121666, _p - 2, _p)) % _p
_sqrt_m1 = pow(2, (_p - 1) // 4, _p)


class WeakKeyError(ValueError):
    """Raised when an Ed25519 public key is non-canonical, off-curve, or small-order."""


def _recover_x(y: int, sign: int):
    if y >= _p:
        return None  # non-canonical y encoding
    x2 = (y * y - 1) * pow(_d * y * y + 1, _p - 2, _p) % _p
    if x2 == 0:
        return None if sign else 0
    x = pow(x2, (_p + 3) // 8, _p)
    if (x * x - x2) % _p != 0:
        x = x * _sqrt_m1 % _p
    if (x * x - x2) % _p != 0:
        return None  # not a square -> point not on curve
    if (x & 1) != sign:
        x = _p - x
    return x


def _decode(pk: bytes):
    """Decode a 32-byte Ed25519 public key to extended coords, or None if invalid."""
    if len(pk) != 32:
        return None
    y = int.from_bytes(pk, "little")
    sign = (y >> 255) & 1
    y &= (1 << 255) - 1
    x = _recover_x(y, sign)
    if x is None:
        return None
    return (x, y, 1, x * y % _p)


def _add(P, Q):
    A = (P[1] - P[0]) * (Q[1] - Q[0]) % _p
    B = (P[1] + P[0]) * (Q[1] + Q[0]) % _p
    C = 2 * P[3] * Q[3] * _d % _p
    D = 2 * P[2] * Q[2] % _p
    E, F, G, H = B - A, D - C, D + C, B + A
    return (E * F % _p, G * H % _p, F * G % _p, E * H % _p)


def _mul8(P):
    return _add(_add(_add(P, P), _add(P, P)), _add(_add(P, P), _add(P, P)))


def _is_identity(P) -> bool:
    # affine (0, 1): X/Z == 0 and Y/Z == 1  <=>  X == 0 and Y == Z (mod p)
    return P[0] % _p == 0 and (P[1] - P[2]) % _p == 0


def is_small_order(pk: bytes) -> bool:
    """True iff pk decodes to a point whose order divides 8 (incl. identity).

    Returns False for keys that fail to decode (handled by check_ed25519_public_key).
    """
    P = _decode(pk)
    if P is None:
        return False
    return _is_identity(_mul8(P))


def check_ed25519_public_key(pk: bytes) -> None:
    """Raise WeakKeyError if pk is non-canonical, off-curve, or small-order.

    Returns None for a canonical, on-curve, large-order public key.
    """
    P = _decode(pk)
    if P is None:
        raise WeakKeyError("public key is not a canonical on-curve Ed25519 point")
    if _is_identity(_mul8(P)):
        raise WeakKeyError("public key is a small-order Ed25519 point")


__all__ = ["WeakKeyError", "check_ed25519_public_key", "is_small_order"]
